### ISSUE TITLE
'Context' in Agent Tasks 

### ISSUE THREAD

**Issue by liuxiaojiao** [joaomdmoura/crewAI#759]
In Agent development, if you write the Tasks in a class with tasks functions as methods. 

```
class ....():
    def task3(self, agent):
        return Task(description='''.....'''),
            expected_output=("...."),
            **context=[self.task1, self.task2],**
            agent=agent) 
```
There is an error on 'context' if you want to use results from previous tasks. 

> pydantic_core._pydantic_core.ValidationError: 2 validation errors for Task
> context.0
>   Input should be a valid dictionary or instance of Task [type=model_type, input_value=<bound method Freelancer_...ject at 0x7fbc6a372fe0>>, input_type=method]
>     For further information visit https://errors.pydantic.dev/2.7/v/model_type

It looks like you have to feed the instances instead of the methods to the context in task 3 by calling task1 & task2 first.

**Comment by lugq1990** [comment]
you are right, for now if the agent try to execute, then context should be the real string type.

related to the source code of the agent execute:
```python
def execute(
        self,
        agent: Union[Agent, None] = None,
        context: Optional[str] = None,   ===> should be string
        tools: Optional[List[Any]] = None,
    ) -> str
```

**Comment by greg80303** [comment]
@liuxiaojiao While I agree that this is poorly documented,  the `context` field in the Task constructor is only used when you have async tasks (`async_execution=True`).  Normally Tasks are always executed in a linear fashion (regardless of `Process.sequential` or `Process.hierarchical`).  But when you have async tasks, they are executed in separate threads and their contexts are gathered (concatenated) as input to the Task that has them specified in its `context` field.

```mermaid
graph TB
    task3((Task 3))
    task4((Task 4))
    task5((Task 5))
    task1((Task 1)) --> task2((Task 2)) --> task3 & task4 & task5 --> task6((Task6))
```

The below example has all other constructor parameters purposely omitted -- just showing the ones discussed in my answer
```python
task1 = Task()
task2 = Task()
task3 = Task(async_execution=True)
task4 = Task(async_execution=True)
task5 = Task(async_execution=True)
task6 = Task(context=[task3, task4, task5])
```

**Comment by strnad** [comment]
> @liuxiaojiao While I agree that this is poorly documented, the `context` field in the Task constructor is only used when you have async tasks (`async_execution=True`). Normally Tasks are always executed in a linear fashion (regardless of `Process.sequential` or `Process.hierarchical`). But when you have async tasks, they are executed in separate threads and their contexts are gathered (concatenated) as input to the Task that has them specified in its `context` field.
> 
> ```mermaid
> graph TB
>     task3((Task 3))
>     task4((Task 4))
>     task5((Task 5))
>     task1((Task 1)) --> task2((Task 2)) --> task3 & task4 & task5 --> task6((Task6))
> ```
> The below example has all other constructor parameters purposely omitted -- just showing the ones discussed in my answer
> 
> ```python
> task1 = Task()
> task2 = Task()
> task3 = Task(async_execution=True)
> task4 = Task(async_execution=True)
> task5 = Task(async_execution=True)
> task6 = Task(context=[task3, task4, task5])
> ```

thanks a lot for clarification, the async tasks now works perfectly :))

**Comment by mdhuzaifapatel** [comment]
> @liuxiaojiao While I agree that this is poorly documented, the `context` field in the Task constructor is only used when you have async tasks (`async_execution=True`). Normally Tasks are always executed in a linear fashion (regardless of `Process.sequential` or `Process.hierarchical`). But when you have async tasks, they are executed in separate threads and their contexts are gathered (concatenated) as input to the Task that has them specified in its `context` field.
> 
> ```mermaid
> graph TB
>     task3((Task 3))
>     task4((Task 4))
>     task5((Task 5))
>     task1((Task 1)) --> task2((Task 2)) --> task3 & task4 & task5 --> task6((Task6))
> ```
> The below example has all other constructor parameters purposely omitted -- just showing the ones discussed in my answer
> 
> ```python
> task1 = Task()
> task2 = Task()
> task3 = Task(async_execution=True)
> task4 = Task(async_execution=True)
> task5 = Task(async_execution=True)
> task6 = Task(context=[task3, task4, task5])
> ```

I have only two tasks : Task1 & Task2 
Task 1 -> Collects data about a topic, mostly some numbers and figures.
Task 2 -> Writes a python code to visualize the data using the data from Task 1
This is how it should work.

I tried adding "async_execution=True" for Task1 and passed it as a context to Task2
> ```task1 = Task(async_execution=True)
> task2 = Task(context=[task1])
> ```

Once it worked, Task2 took the data from Task1 and plotted it. 
But next time when I tried, it's not using the data from Task1, instead it is generating the code independently.

So please let me know how I can achieve this? @greg80303 
What to do when there are only two tasks?
